pkg/metrics/queue.go

```diff
@@ -301,12 +301,46 @@ func (m *QueueMetrics) AddReleasedContainers(value int) {
 	m.containerMetrics.WithLabelValues(ContainerReleased).Add(float64(value))
 }
 
-func (m *QueueMetrics) SetQueueGuaranteedResourceMetrics(resourceName string, value float64) {
-	m.setQueueResource(QueueGuaranteed, resourceName, value)
+func (m *QueueMetrics) SetQueueResourceMetrics(state string, resourceName string, value float64) {
+	m.setQueueResource(state, resourceName, value)
 }
 
-func (m *QueueMetrics) SetQueueMaxResourceMetrics(resourceName string, value float64) {
-	m.setQueueResource(QueueMax, resourceName, value)
+func (m *QueueMetrics) SetQueueNilResourceMetrics(state string) {
+	currResource, err := m.GetAllQueueStateMetrics(state)
+	if err != nil {
+		log.Log(log.Metrics).Warn("failed to get current guaranteed resource metrics",
+			zap.Error(err))
+		return
+	}
+	for k := range currResource {
+		m.SetQueueResourceMetrics(state, k, float64(0))
+	}
+}
+
+func (m *QueueMetrics) GetAllQueueStateMetrics(state string) (map[string]float64, error) {
+	metrics := make(map[string]float64)
+	metricFamilies, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, metricFamily := range metricFamilies {
```

Comment:
Can this be simplified? This seems... extreme. Perhaps we can simply keep (as part of the QueueMetrics object) a list of known resource types for the current queue. Then we can pass in all the resources types at once into SetQueueResourceMetrics(). Iterate both old and new types. Old types that don't exist in the new collection get emitted with zero; new types get added to the known list. Values in the new collection get emitted as-is. Be sure to us appropriate locking.

This allows a much more efficient lookup that doesn't involve traversing the entirety of the metrics, and provides a much more convenient interface for callers.

Reply:
Good idea, i will address this in new PR.

Reply:
It seems we have a lock for metric level, but not queue level:
```
func init() {
	once.Do(func() {
		m = &Metrics{
			scheduler: InitSchedulerMetrics(),
			queues:    make(map[string]*QueueMetrics),
			event:     initEventMetrics(),
			lock:      locking.RWMutex{},
			runtime:   initRuntimeMetrics(),
		}
	})
}
```

Do we need to create a queue level lock to lock the knownResourceType?

Reply:
The individual metric instance needs a lock (see review comments).